core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -944,6 +944,7 @@ private[spark] class DAGScheduler(
     val waiter = new JobWaiter[U](this, jobId, partitions.size, resultHandler)
     eventProcessLoop.post(JobSubmitted(
       jobId, rdd, func2, partitions.toArray, callSite, waiter,
+      JobArtifactSet.getActiveOrDefault(sc),
```

Comment:
This will be the SparkContext. So, if multiple clients adds artifacts via `SparkConnect.addArtifacts` before running a job, the `TaskDescription` will have artifacts from those clients.

Meaning that the cached `IsolatedSessionState` in executor side will have the state with mixed artifacts.

Reply:
>  the TaskDescription will have artifacts from those clients.

This wouldn't be the case because `JobArtifactSet#getActiveOrDefault` ([def](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/JobArtifactSet.scala#L123)) first checks for the "active" set before defaulting back to copying from `SparkContext`.

For the flow of how the `active` set is propagated, here is an example for an `execute` operation in SparkConnext:

1. The execution block is wrapped with `SessionHolder#withSession` [here](https://github.com/apache/spark/blob/master/connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectStreamHandler.scala#L53).
2. `SessionHolder#withContextClassLoader` is then called [here](https://github.com/apache/spark/blob/master/connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SessionHolder.scala#L130) which in turn calls `JobArtifactSet#withActive` [here](https://github.com/apache/spark/blob/master/connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SessionHolder.scala#L118) and sets the active set to `SessionHolder#connectJobArtifactSet`
3. The actual `JobArtifactSet` that is used is built up [here](https://github.com/apache/spark/blob/master/connector/connect/server/src/main/scala/org/apache/spark/sql/connect/artifact/SparkConnectArtifactManager.scala#L157) in `SparkConnectArtifactManager#jobArtifactSet`

Since each client has their own `JobArtifactSet` made `active` when executing an operation, the `TaskDescription` would have artifacts specific to that client and subsequently, `IsolatedSessionState`.